python-lib/wordcloud_generator.py

```diff
@@ -0,0 +1,193 @@
+            self.group_names.append(name)
+
+        # Get tokenization languages differently depending on language/subchart settings combinations
+        if self.language_column == None and self.subchart_column == None:
+            self.languages = [self.language]
+        elif self.language_column != None and self.subchart_column != None:
+            languages, subcharts = zip(*self.group_names)
+            self.languages = list(languages)
+            self.subcharts = list(subcharts)
+        elif self.subchart_column != None:
+            self.subcharts = self.group_names
+            self.languages = [self.language] * len(self.subcharts)
+        else:
+            self.languages = self.group_names
+
+        # Tokenize
+        self.docs = [
+            self.tokenizer.tokenize_list(text, language)[0] for text, language in zip(self.texts, self.languages)
+        ]
+        logging.info(f"Tokenization done in {(time() - start):.2f} seconds.")
+
+    def _count_tokens(self):
+        # Count tokens in each group
+        start = time()
+        logging.info("Initializing count")
+        self.counters = []
+        for doc in self.docs:
+            counter = Counter()
+            for token in doc:
+                counter[(token.text)] += 1  # Equivalently, token.lemma_
+            self.counters.append(counter)
+        logging.info("Count successful, aggregating counters according to chart settings")
+
+        if self.subchart_column == None:
+            # sum the values with same keys
+            self.counts = Counter()
+            for d in self.counters:
+                self.counts.update(d)
+
+            self.counts = dict(self.counts)
+        else:
+            self.counts_df = pd.DataFrame(list(zip(self.subcharts, self.counters)), columns=["subchart", "count"],)
+            self.counts_df = self.counts_df.groupby(by=["subchart"]).agg({"count": "sum"})
+            # remove subcharts emptied by filter
+            self.counts_df = self.counts_df.loc[self.counts_df["count"] != {}, :]
+
+        logging.info(f"Counter aggregation successful, counting done in {(time() - start):.2f} seconds.")
+
+    def _generate_wordclouds(self):
+        # Generate wordclouds and save them as png images
+        start = time()
+        logging.info("Generating wordclouds")
+        if self.subchart_column != None:
+            for name, row in self.counts_df.iterrows():
+                # Generate chart
+                plt.close()
+                wc = self._get_wordcloud(row["count"])
+                fig = plt.figure(figsize=(38.4, 21.6), dpi=100)
+                fig.tight_layout()
+                plt.axis("off")
+                plt.imshow(wc, interpolation="bilinear")
+                # Save chart
+                temp = dkuio.new_bytesoriented_io()
+                fig.savefig(temp, bbox_inches="tight", pad_inches=0, dpi=fig.dpi)
+                self.output_folder.upload_data("wordcloud_" + name + ".png", temp.getvalue())
```

Comment:
You should also pass the name of the category as title to the chart to help with exports

Reply:
The name should be normalized to avoid weird errors it the name contains "/" , ".." or any special character